gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemImpl.java

```diff
@@ -985,18 +996,26 @@ public FileInfo getFileInfoObject(URI path) throws IOException {
     return fileInfo;
   }
 
+  private GoogleCloudStorageItemInfo getFileInfoInternal(
+      StorageResourceId resourceId, boolean inferImplicitDirectories) throws IOException {
+    return getFileInfoInternal(resourceId, inferImplicitDirectories, PathTypeHint.NONE);
+  }
+
   /**
    * @see #getFileInfo(URI)
    */
   private GoogleCloudStorageItemInfo getFileInfoInternal(
-      StorageResourceId resourceId, boolean inferImplicitDirectories) throws IOException {
+      StorageResourceId resourceId, boolean inferImplicitDirectories, PathTypeHint pathTypeHint)
+      throws IOException {
     if (resourceId.isRoot() || resourceId.isBucket()) {
       return gcs.getItemInfo(resourceId);
     }
-    StorageResourceId dirId = resourceId.toDirectoryId();
 
+    StorageResourceId dirId = resourceId.toDirectoryId();
     Future<List<GoogleCloudStorageItemInfo>> listDirFuture =
-        (options.isStatusParallelEnabled() ? cachedExecutor : lazyExecutor)
+        (options.isStatusParallelEnabled() && pathTypeHint != PathTypeHint.FILE
```

Comment:
Shouldn't we log the cases where provided hint is wrong?

Reply:
Done.